pkg/controller/manager.go

```diff
@@ -41,17 +41,22 @@ func GetGlobalStatus() models.ControllerStatuses {
 	return globalStatus.GetStatusModel()
 }
 
-// UpdateController installs or updates a controller in the manager. A
-// controller is identified by its name. If a controller with the name already
-// exists, the controller will be shut down and replaced with the provided
-// controller. Updating a controller will cause the DoFunc to be run
-// immediately regardless of any previous conditions. It will also cause any
-// statistics to be reset.
-func (m *Manager) UpdateController(name string, params ControllerParams) {
-	m.updateController(name, params)
+// UpdateController installs or updates a controller in the
+// manager. A controller is primarily identified by its name.
+// If a controller with the name already exists, the controller
+// will be shut down and replaced with the provided controller.
+//
+// A controller's group is used for aggregate metrics collection.
+// The group.Name must NOT be dynamically generated from a
+// resource identifier, in order to limit metrics cardinality.
+//
+// Updating a controller will cause the DoFunc to be run immediately regardless
+// of any previous conditions. It will also cause any statistics to be reset.
+func (m *Manager) UpdateController(group ControllerGroup, name string, params ControllerParams) {
```

Comment:
Rather than changing the signature of `UpdateController`, what about adding the option to set a group for an entire Manager?

If that doesn't make sense / needs to be overridden, then the ControllerGroup should be a member in ControllerParams

Reply:
Two thoughts there.

Having taken a look at all of the controllers, it seems that many of them do not have an associated Manager, so the group would have to be associated with Controller directly.

As for making ControllerGroup part of the ControllerParams, I had a previous iteration of this that did that, but upon implementation it started to not make sense: `ControllerParams` is passed as the sole parameter to `runController`, and it's expected that these parameters can be "updated" between runs. Since Group is expected to be static for the lifetime of the controller, and is not a parameter of the control loop itself, that approach didn't make sense to me.

Reply:
The group name has been moved back to `ControllerParams`. 

Adding it to `Manager` didn't make sense, at least the way that Manager is currently used. There is no "name" currently associated with a Manager, and many different existing groups (with quite varied purposes) are currently managed by the same Manager. So, adding a group name that would cross-cut these various controllers didn't make sense, given the goal of adding these metrics:

1. The metric should be specific enough that an alert can be constructed with associated documentation on how to recover from the specific type of controller failure.
2. The labels on this metric must be a bounded set of strings, and not be dynamically generated based on Endpoint ID or Policy Name

Adding "name" to controller.Manager doesn't work towards these goals, at least with the current semantics implied by its usage.

Reply:
Gotcha, understood.

There definitely are managers that do meet that definition; but if they are few and far between, there's no need to also add that to the manager.